### PR TITLE
create badge registry smart contract

### LINKED ISSUE #5: Create a registry to map SCM tool handles to an account and a badge ID
- the badge registry can get or mint a badge from a wallet address and return the token id
- the badge registry will hold the mapping between SCM handles and the address/token id

### PR DESCRIPTION
**Pull request by AnthonyBuisset**
closes #5

### CODE REVIEW

**Review by ofux — CHANGES_REQUESTED** [review]
Great work! Very clean. However I didn't approve as I think there is a little bug.

File: contracts/onlydust/deathnote/core/badge_registry/library.cairo [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    member address : felt
+    member badge : BadgeInformation
+    member handles : Handles
+end
+
+#
+# Storage
+#
+@storage_var
+func badge_() -> (address : felt):
```
[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo:34]: You're right

File: contracts/onlydust/deathnote/core/badge_registry/library.cairo [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        with_attr error_message("Badge Registry: Unregistered user"):
+            assert user.address = address
+        end
+
+        return (user)
+    end
+
+    func register_github_handle{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        contributor_address : felt, handle : felt
```
[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo:88]: Yes, or user should be renamed contributor. WDYT ?

File: contracts/onlydust/deathnote/core/badge_registry/test_badge_registry.cairo [hunk 3]
```diff
@@ -1 +1,133 @@
...
+}():
+    alloc_locals
+
+    fixture.initialize()
+
+    let TOKEN_ID = Uint256(0, 0)
+
+    %{
+        stop_prank = start_prank(ids.OWNER)
+        mock_call(ids.BADGE, 'mint', [0, 0])
```
[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/test_badge_registry.cairo:118]: Indeed, also I did not find a way to deploy a contract from a given address

File: contracts/onlydust/deathnote/core/badge_registry/library.cairo [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        with_attr error_message("Badge Registry: Missing Badge contract"):
+            assert_not_zero(badge_contract)
+        end
+
+        if badge_contract - user.badge.contract == 0:
+            return ()  # user badge contract is up-to-date, no need to mint
+        end
+
+        # Update user with minted token
+        let (token_id) = IBadge.mint(badge_contract, user.address)
```
[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo:117]: I will look into it

[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo:117]: Indeed it is a nice finding !

File: contracts/onlydust/deathnote/core/badge_registry/library.cairo (hunk 2 above)
[the remark this answers is not in the document]
Comment (ofux) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo:88]: this is a DDD question :) Yes, you're probably right, "contributor" seems more accurate... unless we plan our users to do other things than contributing (for example, is a user winning a starkling challenge called a contributor?).

[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo:88]: yes, I kept user for the time being

File: contracts/onlydust/deathnote/core/badge_registry/library.cairo [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        return badge_.read()
+    end
+
+    func get_user_information{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        address : felt
+    ) -> (user : UserInformation):
+        let (user) = users_.read(address)
+
+        with_attr error_message("Badge Registry: Unregistered user"):
+            assert user.address = address
```
[the remark this answers is not in the document]
Comment (AnthonyBuisset) [reply contracts/onlydust/deathnote/core/badge_registry/library.cairo]: I ended up removing the user.address field
I was using it to check if a user is registered or not
as the storage_var.read returns 0
I am using now the badge_contract field, which should never be 0 if the user is registered

**Review by ofux — APPROVED** [review]

**Pull request merged**